assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -37,6 +90,44 @@ spec:
       - name: kube-apiserver
         image: {{ .Values.apiserver.image }}
         command:
+        {{- if eq (int .Values.apiserver.replicas) 1 }}
```

Comment:
I'd add a big fat comment explaining that we use different things for 1 replica vs more replicas, and explain HAProxy hacks maybe here too?

I think it should be explained somewhere in the code. The code is quite long because of this, the user/devs reading it can use a hint here to understand these big and very different configurations

Reply:
I'm not sure if putting a big comment in already big YAML make a lot of sense, `git` history has the reasoning. Considering, that it takes a lot of lines to describe why we do it, I wouldn't put the copy in the YAML, in the commit message, in the PR and on Slack :) I can add a short reference if you find it valuable.

Reply:
We could perhaps rename this parameter to something like `controllers_count` or something to make it more self-explainatory.

Reply:
I think a reference (here or in other part of the code) is very valuable. It's very important for anyone reading that file, IMHO. It's a non trivial architecture and we need to describe it somehow, IMHO, so people can do reasonable modifications (and not things that will fail because they don't understand the architecture)

Reply:
Added some short explanation.

Reply:
Maybe not pushed? Or where is it? :)

Reply:
```diff
--- a/assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml
+++ b/assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml
@@ -1,5 +1,28 @@
+# On platforms like AWS, it's enough for kube-apiserver to listen on Node InternalIP. On other
+# like Packet, we need to listen on all interfaces to be able to expose kube-apiserver to the
+# internet. As we cannot listen on specific IP address and on 0.0.0.0 on the same port when using
+# SO_REUSEPORT, kube-apiserver process must listen on different port than HAProxy does. So if we
+# need to expose kube-apiserver on all interfaces, kube-apiserver will be listening on random
+# 127.0.0.0/8 IP address on port 7443, but will advertise Node InternalIP with port 7443 and HAProxy
+# will be listening on both Node InternalIP on port 7443 and on 0.0.0.0:6443 for both internal and
+# external traffic.
+# See commit messages and comments :https://github.com/kinvolk/lokomotive/pull/32 for more details.
+{{ define "port" }}
+{{- if and .Values.apiserver.exposeOnAllInterfaces (eq (int .Values.apiserver.replicas) 1) -}}
+7443
+{{- else -}}
+6443
+{{- end -}}
+{{ end }}
 apiVersion: apps/v1
+# If there is just one controller node, we want to use Deployment to be able to run 2 kube-apiserver
+# pods on a single node at a time, to provide graceful upgrades.
+# See commit messages and comments in https://github.com/kinvolk/lokomotive/pull/32 for more details.
+{{- if eq (int .Values.apiserver.replicas) 1 }}
+kind: Deployment
+{{- else }}
 kind: DaemonSet

```

Reply:
Ohh, sorry! Thanks!